File: MultimediaDemo/app/src/main/java/com/example/multimediademo/MainActivity.java
package com.example.multimediademo;

import androidx.appcompat.app.AppCompatActivity;

import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.MediaController;
import android.widget.VideoView;

public class MainActivity extends AppCompatActivity {
    // 미디어플레이어 인스턴스 생성 - 미디어 재생기를 만듦
    MediaPlayer mediaPlayer = new MediaPlayer();
    VideoView videoView = null;     // 비디오를 재생하기 위한 비디오뷰 인스턴스 생성

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        videoView = findViewById(R.id.videoView);   // findViewById를 이용해 화면의 VideoView와 연결
    }

    public void playAudio(View view){
        // To do. 오디오 재생에 관한 코드
        if (mediaPlayer.isPlaying()){   // 현재 미디어 플레이어가 재생 중이면
            mediaPlayer.pause();        // 재생을 멈춤
        } else {                        // 현재 재생 중이 아니면, 즉, 일시 중지 상태면
            // hare.mp3을 탑재한 미디어 재생기
            mediaPlayer = MediaPlayer.create(getApplicationContext(), R.raw.hare);
            mediaPlayer.setLooping(true);   // 무한 재생 옵션 설정
            mediaPlayer.start();            // 재생한다.
        }
    }

    public void playVideo(View view){
        // To do. 비디오 재생에 관한 코드
        if (videoView.isPlaying()){ // 만약 영상이 재생 중이면
            videoView.pause();      // 일시 중지함
        } else {
            Uri uri = Uri.parse("android.resource://"+getPackageName()+"/raw/movie");
            videoView.setVideoURI(uri); // uri 위치에 있는 영상을 VideoView에 설정
            videoView.start();          // 재생 시작
            videoView.setVisibility(View.VISIBLE);  // 뷰화면에 동영상이 보도록 설정
            // 미디어 컨트롤러 - 재생, 중지, 시크바 등 제어 버튼 추가
            MediaController mediaController = new MediaController(this);
            videoView.setMediaController(mediaController); // 비디어뷰의 컨트롤바로 등록
        }
    }

    @Override
    protected void onDestroy() {    // 앱이 종료될 때 실행됨
        super.onDestroy();
        if(mediaPlayer.isPlaying()){
            mediaPlayer.stop();
            mediaPlayer.release();
            mediaPlayer = null;
        }
        
        if(videoView.isPlaying()){
            videoView.pause();
            videoView = null;
        }
    }
}

File: DialogDemo/app/src/main/java/com/example/dialogdemo/MainActivity.kt
package com.example.dialogdemo

import android.app.AlertDialog
import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.Context
import android.content.DialogInterface
import android.media.MediaPlayer
import android.media.RingtoneManager
import android.net.Uri
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.CombinedVibration
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.util.Log
import android.view.LayoutInflater
import android.widget.DatePicker
import android.widget.TimePicker
import android.widget.Toast
import androidx.annotation.RequiresApi

class MainActivity : AppCompatActivity() {
    @RequiresApi(Build.VERSION_CODES.R)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        //showToast()
//        showDatePick()
//        showTimePicker()
//        showDialog2()
        //customDialog()
//        val ringtoneUri: Uri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)
//        val ringtone = RingtoneManager.getRingtone(applicationContext, ringtoneUri)
//        ringtone.play()

//        val player: MediaPlayer = MediaPlayer.create(this, R.raw.hare)
//        player.start()
        // Vibrator
//        val vibrator: Vibrator
//        vibrator = getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
//        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
//            //val effect = VibrationEffect.createOneShot(100L, VibrationEffect.DEFAULT_AMPLITUDE)
//            val effect = VibrationEffect.createWaveform(longArrayOf(500,1000, 500, 2000), intArrayOf(0, 50, 0, 200), -1)
//            vibrator.vibrate(effect)
//        }else{
//            vibrator.vibrate(500)
//        }

        val vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S){
            getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
        }else{
           getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        }


        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val vibrator = getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager

            //val effect = VibrationEffect.createOneShot(100L, VibrationEffect.DEFAULT_AMPLITUDE)
            val effect = VibrationEffect.createWaveform(longArrayOf(500,1000, 500, 2000), intArrayOf(0, 50, 0, 200), -1)
            val combinedVibration = CombinedVibration.createParallel(effect)
            vibrator.vibrate(combinedVibration)
        }else{
            val vibrator = getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
            //val effect = VibrationEffect.createOneShot(100L, VibrationEffect.DEFAULT_AMPLITUDE)
            val effect = VibrationEffect.createWaveform(longArrayOf(500,1000, 500, 2000), intArrayOf(0, 50, 0, 200), -1)
            vibrator.vibrate(effect)

        }





    }

    fun customDialog(){
//        val inflater = getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
//        val rootView = inflater.inflate(R.layout.dialog_input, null)
//        val builder = AlertDialog.Builder(this)
//            builder.setTitle("Input Dialog")
//            builder.setView(rootView)
//            builder.setPositiveButton("확인", null)
//            builder.show()
    }

    fun showDialog2(){
        val items = arrayOf<String>("사과", "복숭아", "수박", "딸기")
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Item Test")
        builder.setIcon(android.R.drawable.ic_dialog_info)
        builder.setMultiChoiceItems(items, booleanArrayOf(true, false, true, false),
            object : DialogInterface.OnMultiChoiceClickListener {
                override fun onClick(dialog: DialogInterface?, which: Int, isChecked: Boolean) {
                    Log.d( "showDiolog2",
                        "${items[which]}이 ${if (isChecked) "선택되었습니다." else "선택 해제되었습니다."}")
                }
            })
        builder.setPositiveButton("Done",null)
        builder.show()

    }

    fun showDialog1(){
        // 다이얼로그 버튼의 이벤트 핸들러 등록
        val eventHandler = object: DialogInterface.OnClickListener{
            override fun onClick(dialog: DialogInterface?, which: Int) {
                if (which == DialogInterface.BUTTON_POSITIVE){
                    Log.d("showDialog1", "positive button click")
                }else if (which == DialogInterface.BUTTON_NEGATIVE){
                    Log.d("showDialog1", "negative button click")
                }
            }
        }
        // 알림창 빌드
        val build = AlertDialog.Builder(this)
        build.setTitle("Test Dialog")
            .setIcon(android.R.drawable.ic_dialog_info)
            .setMessage("정말로 종료하시겠습니까?")
            .setPositiveButton("Yes",eventHandler)
            .setNegativeButton("No", eventHandler)
            .setNeutralButton("More", null) // 처리할 내용이 없다.
            .show()
    }

    fun showTimePicker(){
        TimePickerDialog(this, object : TimePickerDialog.OnTimeSetListener{
            override fun onTimeSet(view: TimePicker?, hourOfDay: Int, minute: Int) {
                Log.d("tae", "hourOfDay:$hourOfDay, minute: $minute")
            }
        }, 14,30,false).show()
    }

    fun showDatePick(){
        DatePickerDialog(this, object : DatePickerDialog.OnDateSetListener{
            override fun onDateSet(view: DatePicker?, year: Int, month: Int, dayOfMonth: Int) {
                Log.d("tae", "year:$year, month: $month, dayOfMonth:$dayOfMonth")
            }
        }, 2023, 7,3).show()
    }

    @RequiresApi(Build.VERSION_CODES.R)
    fun showToast(){
        val toast = Toast.makeText(this, "종료하려면 한 번 더 누르세요.", Toast.LENGTH_SHORT)
        toast.addCallback(object : Toast.Callback(){
            override fun onToastShown() {   // 토스트가 나타날 때
                super.onToastShown()
                Log.d("tae", "toast hidden")
            }

            override fun onToastHidden() {  // 토스트가 사라질 때
                super.onToastHidden()
                Log.d("tae", "toast shown")
            }
        })
        toast.show()
    }
}
